### PR TITLE
Design interface for logging metrics

### PR DESCRIPTION
**Pull request by charmoniumQ**
Accept after `assorted-fixes-2`

This defines the _interface_ for logging, not the _implementation_ for logging. I want to submit and accept the interface first so that the review burden is smaller, and I am going in the right direction initially (comments/changes to this PR could change the implementation).

Reviewers:
- Read `common/logging.hpp` which defines the logging interface (`logger`) and two utilities (`log_coalescer` and `c_gen_guid`). All are documented appropriately.
- Read `common/record_types.hpp`. This is automatically generated by a script `common/record_types.py` for reasons which will become apparent. However, although it should not be modified directly, it can be read independently (it stands on its own), and it is the best reference for how record_types work.
- No need to review `common/record_types.py` in depth, if the output looks good. See if you can guess how to add a new type, supposing you need to do that.
- Read the "stdout backend" in `runtime/main.cpp`. This is merely a proof-of-concept.

The interface should be designed with the actual implementation in mind: I recommend logging into a SQLite backend, where every record-type is a new table. SQLite is very easy to embed into bigger applications (most web-browsers and video-games embed SQLite). It will be much faster stdout or even file IO.

Interface considerations:
- Component thread will never be slowed down just from logging, so the backend should be multithreaded (which the interface allows). Components call `log_coalescer.log` which might call `logger.log_many` which should _just_ add the buffer of logs to a queue and returns. A thread belonging to the backend will wait on this queue and add to the appropriate database.
- We could potentially exploit parallelism between writes to different tables if writer contention is a problem. A lock is acquired per _database_ in SQLite, so if we just do one table per database, we could get some parallelism.
- Bulk logging many records of the same type will be much more efficient. This is permitted in the interface by `logger.log_many` and `log_coalescer.

Note: I try to always use integer IDs instead of string names. Integer IDs are much more compact and faster to check. Then, I have one table which maps string names to integer IDs. This can be conveniently used with a SQL join clause: `SELECT * FROM (names_to_ids INNER JOIN real_data ON names_to_ids.id = real_data.id) WHERE name = "timewarp";`

### CODE REVIEW

File: common/logging.hpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+		template <typename Record>
+		void log_many(std::vector<std::unique_ptr<const Record>>&& rs) {
+			std::vector<std::unique_ptr<const record>> rs2;
+			for (std::unique_ptr<const Record>& r : rs) {
+				rs2.emplace_back(std::move(r));
+			}
+			// I think (hope) the compiler will optimize this copy-and-cast away.
+			// Unfortunately, there is no way to cast every element of a vector all at once according to
+			// https://stackoverflow.com/a/20310090/1078199
+			log_many2(&Record::type_descr, std::move(rs2));
```
Comment (mhuzai) [inline common/logging.hpp:50]: Where does this call end up?

File: common/logging.hpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	 *
+	 */
+	class c_gen_guid : public phonebook::service {
+	public:
+		/**
+		 * @brief Generate a number, unique from other calls to the same namespace/subnamespace/subsubnamepsace.
+		 */
+		std::size_t get(std::size_t namespace_ = 0, std::size_t subnamespace = 0, std::size_t subsubnamespace = 0) {
+			if (guid_starts[namespace_][subnamespace].count(subsubnamespace) == 0) {
+				guid_starts[namespace_][subnamespace][subsubnamespace].store(1);
```
[the remark this answers is not in the document]
Comment (apache8080) [reply common/logging.hpp:80]: That seems to be happening. Nice catch.

File: common/cpu_timer.hpp [hunk 3]
```diff
@@ -12,38 +13,42 @@
...
  */
 static inline std::chrono::nanoseconds
 cpp_clock_gettime(clockid_t clock_id) {
-	struct timespec ts;
-	if (clock_gettime(clock_id, &ts)) {
-		throw std::runtime_error{std::string{"clock_gettime returned "} + strerror(errno)};
-	}
-	return std::chrono::seconds{ts.tv_sec} + std::chrono::nanoseconds{ts.tv_nsec};
+	/* This ensures the compiler won't reorder this function call; Pretend like it has memory side-effects. */
+	asm volatile (""
```
[the remark this answers is not in the document]
Comment (charmoniumQ) [reply common/cpu_timer.hpp:23]: @mhuzai I believe that is possible according to [this answer](https://stackoverflow.com/a/26195437/1078199)
@apache8080 Yes. The `asm` keyword works the same way, and no actual assembly code is used here; just the side effects.

File: common/logging.hpp (hunk 1 above)
Comment (charmoniumQ) [reply common/logging.hpp:50]: `c_metric_logger` is an abstract class, so there will be some class that extends that interface (call it `impl_class`).
- `c_metric_logger::log` calls `impl_class::log2`
- `c_metric_logger::log_many` calls `impl_class::log_many2`

File: common/logging.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (charmoniumQ) [reply common/logging.hpp]: You are correct. I changed it to post-increment so the instance variable still holds the first unused `guid`.

[the remark this answers is not in the document]
Comment (mhuzai) [reply common/logging.hpp]: This is minor but one last thing: why start from 1 and not 0?

[the remark this answers is not in the document]
Comment (charmoniumQ) [reply common/logging.hpp]: See the comments [here](https://github.com/ILLIXR/ILLIXR/pull/23/files#diff-4ec61cd70cc9f3ddcf7ccd2025c248dbR80).

**Comment by charmoniumQ** [comment]
I won't merge this before #24 on which this depends.

**Comment by charmoniumQ** [comment]
Better version to come.

**Pull request closed without merging**